src/lib.rs

```diff
@@ -27,15 +27,111 @@
 //! other crates (like [`tokio-codec`](https://docs.rs/tokio-codec/) or
 //! [`async-bincode`](https://docs.rs/async-bincode)) and instead operates at the level of
 //! [`Sink`](https://docs.rs/futures/0.1/futures/sink/trait.Sink.html)s and
-//! [`Stream`](https://docs.rs/futures/0.15/futures/stream/trait.Stream.html)s. In particular, it
-//! assumes that there exists a `Sink + Stream` transport where it can send `Request`s and receive
-//! `Response`s, or vice-versa for the server side.
+//! [`Stream`](https://docs.rs/futures/0.15/futures/stream/trait.Stream.html)s.
+//!
+//! At its core, `tokio-tower` wraps a type that is `Sink + Stream`. On the client side, the Sink is used to send requests,
+//! and the Stream is used to receive responses (from the server) to those requests. On the server side, the Stream is
+//! used to receive requests, and the Sink is used to send the responses.  
 //!
 //! # Servers and clients
 //!
 //! This crate provides utilities that make writing both clients and servers easier. You'll find
 //! the client helper as `Client` in the protocol module you're working with (e.g.,
 //! [`pipeline::Client`]), and the server helper as `Server` in the same place.
+//!
+//! # Example
+//! ```rust
+//! # use std::pin::Pin;
+//! # use std::boxed::Box;
+//! # use tokio::sync::mpsc;
+//! # use tokio::io::{AsyncWrite, AsyncRead};
+//! # use futures_core::task::{Context, Poll};
+//! # use futures_util::{never::Never, future::{poll_fn, ready, Ready}};
+//! # use tokio_tower::pipeline;
+//! # use core::fmt::Debug;
+//! type StdError = Box<dyn std::error::Error + Send + Sync + 'static>;
+//!
+//! /// Wrapper around our mpsc channel transport.
+//! ///
+//! /// tokio-tower requires our mpsc-based transport implement Sink + Stream (see "Transports" above), but our mpsc channel operates on
+//! /// items of type T. To bridge this gap, we have to provide implementations of Sink and Stream.
```

Comment:
This reads strangely to me. mpsc channels are already `Sink` (the tx side) and `Stream` (the rx side). The issue isn't the `T` type — the issue is that you need a transport that is _both_ `Sink` _and_ `Stream`.

Reply:
mpsc::Sender is not technically `Sink` (https://docs.rs/tokio/0.2.11/tokio/sync/mpsc/struct.Sender.html), but it might as well be and I see what you are saying. Will reword.